File: lhs712/Assignment 1_ Regular_Expressions/sample_solution.R
file_path = "./dates.txt"
output_file_path = "./output.txt"
sink(output_file_path)

data = read.table(file_path, header = F, sep = '\t', stringsAsFactors=FALSE, quote = "")

# You could define month names as a data frame and use it to normalize, using month_names[month][1,1]
month_names <- data.frame(
  Jan = '01', Feb = '02', Mar = '03', Apr = '04',May = '05',Jun = '06',
  Jul = '07', Aug = '08' , Sep = '09', Oct = '10', Nov = '11', Dec = '12'
)

# Example of how to write a function. This function matches a given pattern in a given line
find_pattern<-function(pattern,line){
  return (regmatches(line, regexpr(pattern, line)));
}

# Example of a pattern finding function
find_pattern_1<-function(line,id){
  pattern = '\\d'
  pattern = "?:\\d{2}"
  pattern = "\\d{1,2}"
  s = find_pattern(pattern, line)
  if(length(s) > 0){
	# Work with the matched section s
    cat(id, "\t", s, "\n", sep="")
    return (TRUE)
  }
  return (FALSE)
}

for(i in 1:500){
  line = data[i,2]
  
  find_pattern_1(line,i)
  # if find_pattern_1 returns FALSE, find another pattern, and so on...
}
sink()

File: lhs712/Assignment 1_ Regular_Expressions/sample_solution.py
import re
import datetime


new_date = datetime.datetime(2000, 1, 1)
pushed_40_date = new_date + datetime.timedelta(days=40)
print(pushed_40_date.strftime("%Y-%m-%d"))
# month_names = {"Jan":"01", "Feb":"02", "Mar":"03", "Apr":"04", "May":"05","Jun":"06", "Jul":"07", "Aug":"08", "Sep":"09", "Oct":"10", "Nov":"11", "Dec":"12"}
# text = 'Jun 1987blaaaaa'
# regx1 = r"(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec).*(\d{4})"
# match1 = re.search(regx1, text)
# if match1:
# 	text = re.sub(regx1, f"{month_names[match1.group(1)]}/{match1.group(2)}", text)

# print(text)
# def find_dates(filename):
# # You could define the following to normalize months, using month_names[month]
# 	month_names = {"Jan":"01", "Feb":"02", "Mar":"03", "Apr":"04", "May":"05","Jun":"06", "Jul":"07", "Aug":"08", "Sep":"09", "Oct":"10", "Nov":"11", "Dec":"12"}

# 	with open(filename, 'r') as file:
# 		for line in file:
# 			match = re.search(r"(\d+)\t(.*)", line)
# 			id = match.group(1)
# 			text = match.group(2)

# 			# Search regular expressions in text. For example, the line below searches for a digit
# 			match = re.search(r"(May|Feb)", text)
# 			if match:
# 				# work with what matched. match.group(0) is everthing that matched the regex. If you used () in your regex, match.group(1) represents the first (), and so on...
# 				print (id + "\t" + match.group(0)) ;
# # Jan 1987
# # Marc, 1981
# # June, 1999
# # January 2007

# # June 25, 2012
# # September. 15, 2011
# # 18 Jan 1990
# (Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec).*(\d{4})


# if __name__ == '__main__':
# 	find_dates('./dates.txt')
